File: py/kintsugi/imports.py
"""A rough Python import graph, built by scanning statements rather than
parsing.

Its only job is blast radius: how many modules would a change to this file
reach? It never needs to be exact — a heuristic that overcounts importers
escalates a few extra files, and undercounting is the risk, so it errs
toward treating a file as shared. This is the Python half of the TS engine's
imports.ts (TS/JS scanning stays with the Node engine).
"""

import glob
import os
import re

SKIP = re.compile(r"(^|[\\/])(node_modules|dist|build|\.git)([\\/]|$)")
PY_IMPORT_RE = re.compile(r"^\s*(?:from\s+([.\w]+)\s+import|import\s+([.\w]+))", re.MULTILINE)
# Test files are not product importers — they exercise the code, not own it.
TEST_FILE = re.compile(r"(?:^|[/\\])(?:test_|_test)\.py$")


def build_import_graph(source_root: str):
    importers = {}
    deps = {}
    files = []
    for f in glob.glob(os.path.join(source_root, "**", "*.py"), recursive=True):
        rel = os.path.relpath(f, source_root).replace("\\", "/")
        if SKIP.search(rel):
            continue
        files.append(f)
    file_set = set(files)

    for file in files:
        with open(file, "r", encoding="utf-8", errors="replace", newline="") as fh:
            text = fh.read()
        targets = set()
        for m in PY_IMPORT_RE.finditer(text):
            spec = (m.group(1) or m.group(2) or "").strip()
            if not spec:
                continue
            target = _resolve_spec(spec, file, source_root, file_set)
            if target:
                targets.add(target)
        deps[file] = targets
        # A file imported by a *test* is not a shared product module — tests
        # reference modules they exercise, and editing those modules is
        # normal. Only product-code importers count toward blast radius.
        if TEST_FILE.search(file):
            continue
        for t in targets:
            importers.setdefault(t, set()).add(file)

    return {"importers": importers, "deps": deps, "files": files}


def _resolve_spec(spec: str, file: str, source_root: str, file_set: set) -> str | None:
    if spec.startswith("."):
        dots = len(re.match(r"^\.+", spec).group(0))
        rest = spec[dots:]
        parts = [".."] * (dots - 1) + ([rest] if rest else [""])
        base = os.path.normpath(os.path.join(os.path.dirname(file), *parts))
        candidates = [f"{base}.py", os.path.join(base, "__init__.py")]
    else:
        base = os.path.join(source_root, *spec.split("."))
        candidates = [os.path.join(base, "__init__.py")]
    for cand in candidates:
        if cand in file_set:
            return cand
    return None


def scope_of(graph, file: str):
    """Blast radius from the shape of the graph, not from the patch. A file
    imported by two or more modules is shared: editing it moves code the loop
    was not looking at, so the repair is escalated instead of applied."""
    n = len(graph["importers"].get(file, set()))
    return ("shared" if n >= 2 else "local"), n

File: src/imports.ts
import { readFileSync, globSync } from 'node:fs';
import { resolve, dirname } from 'node:path';

/**
 * A rough import graph, built by scanning statements rather than parsing.
 *
 * Its only job is blast radius: how many modules would a change to this file
 * reach? It never needs to be exact — a heuristic that overcounts importers
 * escalates a few extra files, and undercounting is the risk, so the rules
 * below err toward treating a file as shared.
 */

const SRC_GLOB = '**/*.{ts,tsx,js,jsx,mjs,cjs,py}';
const SKIP = /(^|[\\/])(node_modules|dist|build|\.git)([\\/]|$)/;
const IMPORT_RE = /(?:import|export)\s+(?:[^'"]*?\s+from\s+)?['"]([^'"]+)['"]|require\(['"]([^'"]+)['"]\)/g;
// Python: `from x import y` / `import x` — the module spec, relative ones
// included (`from ..pkg import`, `from . import`). Non-relative specs are
// resolved against the source root (first-party packages), so blast radius
// is real for .py files too.
const PY_IMPORT_RE = /^\s*(?:from\s+([.\w]+)\s+import|import\s+([.\w]+))/gm;
/** Test files are not product importers — they exercise the code, not own it. */
const TEST_FILE = /\.(test|spec)\.[cm]?[jt]sx?$|(?:^|[/\\])test_[^/\\]*\.py$|(?:^|[/\\])[^/\\]*_test\.py$/;

export interface ImportGraph {
  /** absolute file -> absolute files that import it */
  importers: Map<string, Set<string>>;
  /** absolute file -> absolute files it imports (resolved, in-tree) */
  deps: Map<string, Set<string>>;
  files: string[];
}

export function buildImportGraph(sourceRoot: string): ImportGraph {
  const importers = new Map<string, Set<string>>();
  const deps = new Map<string, Set<string>>();
  const files: string[] = [];

  for (const f of globSync(SRC_GLOB, { cwd: sourceRoot })) {
    const rel = f.replace(/\\/g, '/');
    if (SKIP.test(rel)) continue;
    files.push(resolve(sourceRoot, f));
  }
  const fileSet = new Set(files);

  for (const file of files) {
    const text = readFileSync(file, 'utf8');
    const set = new Set<string>();
    if (file.endsWith('.py')) {
      for (const m of text.matchAll(PY_IMPORT_RE)) {
        const spec = (m[1] ?? m[2] ?? '').trim();
        if (!spec) continue;
        let base: string;
        let rest = '';
        if (spec.startsWith('.')) {
          const dots = spec.match(/^\.+/)?.[0].length ?? 0;
          rest = spec.slice(dots);
          base = resolve(dirname(file), ...Array(dots - 1).fill('..'), ...(rest ? [rest] : ['']));
        } else {
          base = resolve(sourceRoot, ...spec.split('.'));
        }
        const cands = rest
          ? [`${base}.py`, resolve(base, '__init__.py')]
          : [resolve(base, '__init__.py')];
        for (const cand of cands) {
          if (fileSet.has(cand)) { set.add(cand); break; }
        }
      }
    } else {
      for (const m of text.matchAll(IMPORT_RE)) {
        const spec = m[1] ?? m[2];
        if (!spec || !spec.startsWith('.')) continue;
        const base = resolve(dirname(file), spec);
        let target: string | null = null;
        // NodeNext projects import './pricing.js' for a file that is actually
        // pricing.ts, so runtime extensions must map back to source files.
        const mapped = [
          base,
          `${base}.ts`, `${base}.tsx`, `${base}.js`, `${base}.jsx`, `${base}.mjs`,
          `${base}/index.ts`, `${base}/index.js`,
          base.replace(/\.js$/, '.ts'), base.replace(/\.jsx$/, '.tsx'),
          base.replace(/\.mjs$/, '.mts'), base.replace(/\.cjs$/, '.cts'),
        ];
        for (const cand of mapped) {
          if (fileSet.has(cand)) { target = cand; break; }
        }
        if (target && fileSet.has(target)) set.add(target);
      }
    }
    deps.set(file, set);
    // A file imported by a *test* is not a shared product module — tests
    // reference modules they exercise, and editing those modules is normal.
    // Only product-code importers count toward blast radius.
    if (TEST_FILE.test(file)) continue;
    for (const t of set) {
      if (!importers.has(t)) importers.set(t, new Set());
      importers.get(t)!.add(file);
    }
  }

  return { importers, deps, files };
}

/**
 * Blast radius from the shape of the graph, not from the patch. A file
 * imported by two or more modules is shared: editing it moves code the loop
 * was not looking at, so the repair is escalated instead of applied.
 */
export function scopeOf(
  graph: ImportGraph,
  file: string,
): { scope: 'local' | 'shared'; importers: number } {
  const n = graph.importers.get(file)?.size ?? 0;
  return { scope: n >= 2 ? 'shared' : 'local', importers: n };
}
